# Project: Credit Card Fraud Detection

## Table of Contents

1. [Importing Libraries](#section-one)
2. [Loading the raw data](#section-two)
3. [Data Pre-Processing](#section-three)
4. [Model Training](#section-four)
5. [Model Evaluation](#section-five)

<a id="section-one"></a>
## 1. Importing Libraries 

In [1]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression

<a id="section-two"></a>
## 2. Loading the raw data 

In [2]:
# Panda Dataframe

raw_data = pd.read_csv('C:\\Users\\Nishant Gupta\\Desktop\\Projects\\classification_project_Nishant_Gupta\\creditcard.csv')

# Let's have a look at the data

raw_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

So we have the first column as the transaction time. The columns V1 to V28 are the features related to the credit card. The companies cannot provide the names of features as it comes under sensitive information! Then column 'Amount' gives the credit card transactions in US dollars. Finally, we have the column 'Class' which tells us whether the transaction is legit or not so it only has two values: 1 and 0.

### Information about the dataset 

In [3]:
# let's check the number of rows and columns in the dataset

raw_data.shape

(284807, 31)

We have 284807 rows and 31 columns in this dataset. 

In [4]:
# Some information about the dataset

raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [5]:
# Let's have a look at the descriptive statistics of the dataset

raw_data.describe(include='all')

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  3.918649e-15  5.682686e-16 -8.761736e-15  2.811118e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean  -1.552103e-15  2.040130e-15 -1.698953e-15 -1.893285e-16 -3.147640e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.473120e-16  8.042109e-16  5.282512e-16  4.456271e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   1.426896e-15  1.701640e-15 -3.662252e-16 -1.217809e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [6]:
# Checking the total number of missing values

raw_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

Great! There are no missing values, thus we won't have to eliminate any variables.

In [7]:
# We have checked the raw data and there aren't any missing values!

data_no_mv = raw_data.copy()

In [8]:
# Checking the number of legit transactions and fraudulent transactions

data_no_mv['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

The dataset is very unbalanced as major data is fraudulent.

In [9]:
# Seperating the data of legit and fraud transactions 

legit_trans = data_no_mv[data_no_mv.Class == 0]
fraud_trans = data_no_mv[data_no_mv.Class == 1]

# Let's have a look at the shape

print(legit_trans.shape)
print(fraud_trans.shape)

(284315, 31)
(492, 31)


In [10]:
# Let's see the statistical measures of this data

legit_trans.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [11]:
legit_trans.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [12]:
# compare the value for both transactions

data_no_mv.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

<a id="section-three"></a>
## 3. Data Pre-Processing 

### Splitting the Features and Targets 

In [13]:
y = data_no_mv['Class']
x = data_no_mv.drop(columns = 'Class', axis = 1)

In [14]:
y.value_counts()

0    284315
1       492
Name: Class, dtype: int64

### SMOTE 

In [15]:
from imblearn.over_sampling import SMOTE

smote = SMOTE(sampling_strategy='minority')
x_sm, y_sm = smote.fit_resample(x, y)

y_sm.value_counts()

0    284315
1    284315
Name: Class, dtype: int64

In [16]:
# Let's have a look

print(x_sm)

                Time         V1         V2         V3        V4         V5  \
0           0.000000  -1.359807  -0.072781   2.536347  1.378155  -0.338321   
1           0.000000   1.191857   0.266151   0.166480  0.448154   0.060018   
2           1.000000  -1.358354  -1.340163   1.773209  0.379780  -0.503198   
3           1.000000  -0.966272  -0.185226   1.792993 -0.863291  -0.010309   
4           2.000000  -1.158233   0.877737   1.548718  0.403034  -0.407193   
...              ...        ...        ...        ...       ...        ...   
568625  70162.307131  -0.003360   1.425919  -2.240368  3.696678  -1.544405   
568626  28200.860390 -27.298788  15.423229 -28.526663  6.379512 -20.166312   
568627  41285.275893 -12.180435   6.481584 -13.357924  8.077400 -11.116097   
568628  47768.954817   0.695427   1.082278  -1.515842  1.515122  -0.631370   
568629  42266.685005  -9.699966   5.595180 -10.454323  6.143040  -3.360200   

              V6         V7         V8        V9  ...       V20

In [17]:
print(y_sm)

0         0
1         0
2         0
3         0
4         0
         ..
568625    1
568626    1
568627    1
568628    1
568629    1
Name: Class, Length: 568630, dtype: int64


### Splitting the data into training data & testing data 

In [18]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x_sm, y_sm, test_size = 0.25, stratify = y_sm, random_state = 101)

In [19]:
# By looking at the shape, we could see the number of observations which are training and testing

print(x.shape, x_train.shape, x_test.shape)

(284807, 30) (426472, 30) (142158, 30)


In [20]:
# Feature Scaling

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

x_train = scaler.fit_transform(x_train)

x_test = scaler.fit_transform(x_test)

<a id="section-four"></a>
## 4. Model Training 

### Logistic Regression 

In [21]:
model = LogisticRegression(max_iter=10000)

In [22]:
# training the Logistic Regression Model with Training Data

model.fit(x_train, y_train)

LogisticRegression(max_iter=10000)

<a id="section-five"></a>
## 5. Model Evaluation 

In [23]:
# Accuracy Score
# we will find the accuracy on training data

from sklearn.metrics import accuracy_score

x_train_pred = model.predict(x_train)

train_data_accuracy = accuracy_score(x_train_pred, y_train)

In [24]:
print('Accuracy on training data: ', train_data_accuracy)

Accuracy on training data:  0.9804629612260594


In [25]:
# we will find accuracy on test data

x_test_pred = model.predict(x_test)

test_data_accuracy = accuracy_score(x_test_pred, y_test)

In [26]:
print('Accuracy on testing data: ', test_data_accuracy)

Accuracy on testing data:  0.980359881258881


Accuracy is important to understand if model is deployed well. If the accuracy score of training data is very different from testing data then the model is most likely overfitted or underfitted.

### Confusion Matrix

In [27]:
# creating the confusion matrix

from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, x_test_pred)

array([[70423,   656],
       [ 2136, 68943]], dtype=int64)

In [28]:
# extracting true negative, true positive, false positive, false negative

tn, fp, fn, tp = confusion_matrix(y_test, x_test_pred).ravel()

(tn, fp, fn, tp)

(70423, 656, 2136, 68943)

In [29]:
# Confusion matrix metrics

from sklearn.metrics import classification_report

matrix = classification_report(y_test, x_test_pred)

print('Classification report : \n',matrix)

Classification report : 
               precision    recall  f1-score   support

           0       0.97      0.99      0.98     71079
           1       0.99      0.97      0.98     71079

    accuracy                           0.98    142158
   macro avg       0.98      0.98      0.98    142158
weighted avg       0.98      0.98      0.98    142158

